OpenRA.Game/GameRules/RulesetWatcher.cs

```diff
@@ -0,0 +1,172 @@
+				watchFiles[fullPath] = file;
+			}
+
+			timer = new Timer(10) { AutoReset = false };
+			timer.Elapsed += RefreshTimer;
+
+			watcher = new FileSystemWatcher(modData.Manifest.Package.Name)
+			{
+				IncludeSubdirectories = true
+			};
+			watcher.Changed += FileChanged;
+
+			foreach (var file in watchFiles.Keys.Distinct())
+			{
+				watcher.Filters.Add(Path.GetFileName(file));
+			}
+		}
+
+		public void ToggleWatching(bool toggleValue)
+		{
+			if (isDisposed)
+				throw new ObjectDisposedException(nameof(RulesetWatcher));
+
+			watcher.EnableRaisingEvents = toggleValue;
+			isEnabled = toggleValue;
+			if (toggleValue)
+				RestartTimerIfEnabled();
+			else
+			{
+				timer.Stop();
+
+				lock (fileQueue)
+					fileQueue.Clear();
+			}
+		}
+
+		void RefreshTimer(object sender, ElapsedEventArgs e)
+		{
+			if (fileQueue.Count == 0)
+			{
+				RestartTimerIfEnabled();
+				return;
+			}
+
+			List<string> copy;
+			lock (fileQueue)
+			{
+				if (fileQueue.Count == 0)
+					return;
+
+				copy = fileQueue.ToList();
+				fileQueue.Clear();
+			}
+
+			if (isDisposed || !isEnabled)
+				return;
+
+			Game.RunAfterTick(() =>
+			{
+				if (isDisposed || !isEnabled)
+					return;
+
+				var modFsFilenames = copy.Select(f => watchFiles[f]).ToArray();
+
+				var defaultRules = world.Map.Rules;
+				var rulesFiles = FindRulesetFiles(modData.Manifest.Rules, modFsFilenames).ToArray();
+				var weaponFiles = FindRulesetFiles(modData.Manifest.Weapons, modFsFilenames).ToArray();
+				var sequenceFiles = FindRulesetFiles(modData.Manifest.Sequences, modFsFilenames).ToArray();
+
+				if (rulesFiles.Length > 0)
+					TryExecute(() => defaultRules.LoadActorTraitsFromRuleFile(world, modData, rulesFiles));
+				if (weaponFiles.Length > 0)
+					TryExecute(() => defaultRules.LoadWeaponsFromFile(world, modData, weaponFiles));
+				if (sequenceFiles.Length > 0)
+					TryExecute(() => world.Map.Sequences.ReloadSequenceSetFromFiles(modData.DefaultFileSystem, sequenceFiles));
+
+				RestartTimerIfEnabled();
+
+				IEnumerable<string> FindRulesetFiles(IEnumerable<string> allFiles, IEnumerable<string> findFiles)
+				{
+					return allFiles.Where(f => findFiles.Any(c => f.Equals(c, StringComparison.OrdinalIgnoreCase))
+						&& modData.DefaultFileSystem.Exists(f));
+				}
+
+				static void TryExecute(Action action, int attempts = 2)
+				{
+					for (var i = 0; i < attempts; i++)
+					{
+						try
+						{
+							action();
+						}
+						catch (IOException)
+						{
+							// IO errors are not critical, we should be able to retry them a few times (and even swallow them, if they keep occurring)
+							// TODO: error handling should be in each respective code that reloads ruleset/sequences/etc.
+						}
+					}
+				}
+			});
+		}
+
+		void RestartTimerIfEnabled()
+		{
+			if (isEnabled)
+				timer.Start();
+		}
+
+		void FileChanged(object sender, FileSystemEventArgs e)
+		{
+			if (!watchFiles.ContainsKey(e.FullPath))
```

Comment:
This depends on what @AspectInteractive's solution for reloading ruleset can handle. AFAIK it needs to work with rules files defined in the mod manifest.

Reply:
That's right currently it uses files from Manifest only, and I think this is adequate for the first version and would suit 99% of use cases for hot reload (adjusting values rather than creating brand new files). Later we can look at loading any file, however it would need to go through Manifest first to be consistent, so it will be some work to tweak Manifest to enable that capability.